audtorch/datasets/speech_commands.py

```diff
@@ -0,0 +1,117 @@
+            Root Location: /data/speech_commands_v0.02
+            Sampling Rate: 16000Hz
+        >>> signal, target = data[4]
+        >>> target
+        'right'
+        >>> sd.play(signal.transpose(), data.sampling_rate)
+    """
+
+    url = ('http://download.tensorflow.org/'
+           'data/speech_commands_v0.02.tar.gz')
+
+    # Available target commands
+    commands = [
+        'right', 'eight', 'cat', 'tree', 'backward',
+        'learn', 'bed', 'happy', 'go', 'dog', 'no',
+        'wow', 'follow', 'nine', 'left', 'stop', 'three',
+        'sheila', 'one', 'bird', 'zero', 'seven', 'up',
+        'visual', 'marvin', 'two', 'house', 'down', 'six',
+        'yes', 'on', 'five', 'forward', 'off', 'four']
+
+    background_noise = '_background_noise_'
+
+    def __init__(self, root, train=True, download=False, *,
+                 sampling_rate=16000, include=None, silence=True,
+                 transform=None, target_transform=None):
+
+        if download:
+            self.root = safe_path(root)
+            self._download()
+
+        if include is None:
+            include = self.commands
+
+        with open(safe_path(os.path.join(root, 'testing_list.txt'))) as f:
+            test_files = f.read().splitlines()
+
+        files, targets = [], []
+        for speech_cmd in self.commands:
+            d = os.listdir(os.path.join(root, speech_cmd))
+            d = [os.path.join(speech_cmd, x) for x in d]
+
+            # Filter out test / train files using `testing_list.txt`
+            d_f = list(set(d) - set(test_files)) \
+                if train else list(set(d) & set(test_files))
+
+            files.extend([os.path.join(root, p) for p in d_f])
+            target = speech_cmd if speech_cmd in include else 'unknown'
+            targets.extend([target for _ in range(len(d_f))])
+
+        # Match occurrences of silence with `unknown`
+        if silence:
+            n_samples = max(targets.count('unknown'), 3000)
+            n_samples = int(n_samples * 0.9) \
+                if train else int(n_samples * 0.1)
+
+            sf = []
+            for file in os.listdir(os.path.join(root, '_background_noise_')):
+                if file.endswith('.wav'):
+                    sf.append(os.path.join(root, '_background_noise_', file))
+
+            targets.extend(['silence' for _ in range(n_samples)])
+            files.extend(random.choices(sf, k=n_samples))
+
+        super().__init__(root, files, targets, sampling_rate,
+                         transform=transform,
+                         target_transform=target_transform)
+
+    def _download(self):
+        if self._check_exists():
+            return
+        download_dir = self.root
+        corpus = 'speech_commands_v0.02'
+        if download_dir.endswith(corpus):
+            download_dir = download_dir[:-len(corpus)]
+        filename = download_url(self.url, download_dir)
+        extract_archive(filename, remove_finished=True)
```

Comment:
For me, this is not working:
```python
>>> data = SpeechCommands(root='~/tmp/speech_commands_v0.02/', download=True)
                   Downloading http://download.tensorflow.org/data/speech_commands_v0.02.tar.gz to /home/audeering.local/hwierstorf/tmp/speech_commands_v0.02.tar.gz
                                                                                                                                                                                              Extracting /home/audeering.local/hwierstorf/tmp/speech_commands_v0.02.tar.gz
100%|████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████████▉| 2.43G/2.43G [00:52<00:00, 71.5MB/sTraceback (most recent call last):
  File "<stdin>", line 1, in <module>                                                                                                                                                          
  File "/home/audeering.local/hwierstorf/git/audeering/audtorch/audtorch/datasets/speech_commands.py", line 75, in __init__
    with open(safe_path(os.path.join(root, 'testing_list.txt'))) as f:
FileNotFoundError: [Errno 2] No such file or directory: '/home/audeering.local/hwierstorf/tmp/speech_commands_v0.02/testing_list.txt'
```
and
```bash
$ ls ~/tmp
_background_noise_  bed   cat  down   five    forward  go     house  left    nine  off  one    seven   six   three  two  visual  yes   LICENSE    testing_list.txt
backward            bird  dog  eight  follow  four     happy  learn  marvin  no    on   right  sheila  stop  tree   up   wow     zero  README.md  validation_list.txt
```

Why do you do `download_dir = download_dir[:-len(corpus)]`?
Is it working on your machine?

Reply:
The download works for me if I change
```python
extract_archive(filename, remove_finished=True)
```
to
```python
extract_archive(filename, out_path=self.root, remove_finished=True)
```

Reply:
Oops, thanks - the code didn't actually run when I tested it, it caught the _check_exists(). I'll resolve now.

Reply:
> Why do you do download_dir = download_dir[:-len(corpus)]?
I just copied the code in `mozilla_common_voice.py` as an example. 

It means that the download won't be nested (e.g. `~/tmp/speech_cmds/speech_cmds/`) if the root is set to be the same name as the dataset. I've changed it now to something that makes more intuitive sense to me:

     if download_dir.endswith(url.split("/")[-1]):
         download_dir = "/".join(download_dir.split("/")[:-1])

Reply:
Thanks, it is now working for me.